docs/local_assembly.tex

```diff
@@ -0,0 +1,121 @@
+Next, the \code{ReadThreadingAssembler} assembles the (corrected) reads over several different kmer sizes specified by the \code{kmerSize} argument\footnote{By default, 10 and 25.  This size is a compromise and no single value is the best choice for all regions.  Large kmers are more likely to be unique and to yield a graph with no cycles, which is especially important in low-complexity regions, but they are more sensitive to errors and low coverage.}.  If the given kmer sizes fail to produce a graph without cycles, or if more than $1/5$ of the kmers in the graph are not unique in the sequences from which they come\footnote{This uniqueness condition is waived on the last attempt when $k$ has been increased by 60 bases.}, the \code{ReadThreadingAssembler} repeatedly increases $k$ by 10 bases until assembly succeeds or until $k$ has been increased 6 times, that is, by 60 bases\footnote{These constants, 10 and 6, are hard-coded, but the attempts to increase $k$ can be disabled completely with the \code{dontIncreaseKmerSizesForCycles} argument.}.
+
+The first step is to create a set of sequences to be kmerized and put into the de Bruijn graph.  These sequences are: the reference haplotype, any requested alleles given by the \code{alleles} argument in \HC's ``genotype given alleles" (GGA) mode\footnote{This mode is disabled in \Mutect~ and incompatible with the reference confidence mode of \HC, which is the recommended best practice.  We document it here for completeness.}, and maximal subsequences of reads with base quality at least 10 by default\footnote{This is controlled by the \code{minBaseQualityScore} argument.} and at least one kmer long.  That is, a read with 100 bases and a base of quality 7 at position 70 yields sequences of length 69, 30 if $k = 10$ and a single sequence of length 69 if $k = 40$.  We do not use mate information; that is, a read and its mate yield completely independent sequences, nor do we later post-process the graph using mate information.  This is a potential direction for improvement.
+
+Before building the graph, the \code{ReadThreadingAssembler} finds the set of all kmers that appear twice in the same sequence i.e. twice in the reference or twice in any read.  Then each sequence is kmerized starting from the first kmer that is not in this set.  When a new kmer appears, a vertex is added to the graph; otherwise an edge is added (or the multiplicity of an existing edge incremented) from the preceding kmer to the current one.  Up to this point, we have produced a de Bruijn graph in textbook fashion, with the exception that we have excluded kmers containing low-quality bases and non-unique kmers at the beginning of reads.  Note that this is a directed graph of kmers from aligned reads (i.e. aligned to the forward strand of the reference), not a bidirected graph of unaligned reads that could come from either strand, with the complications of associating a kmer with its reverse complement.
```

Comment:
There are a few innacuracies in this paragraph for what I remember, I'm going to addreses those here without confirming looking at the code so there is a risk that I'm wrong and you are right...

> "Before building the graph, the ReadThreadingAssembler funds the set of all kners that appear twice in the same sequence i.e. ref or read". 

I think that you mean to say twice or more, so not once. Also you don't say what those kmers are (the **non-unique kmer set**)... just that you won't use them as the start "threading the sequence"... I think it is important to explicitly state the concepts of unique and non-unique kmers as they have a important influence as how the rest of the algorithm progresses. 

The very first sequence that we thread is the reference sequence. In that special case we add all the kmers to the graph (I believe) unique or non-unique.
I'm not sure what is done with the given alleles in GGA mode... I think that if GGA is on we don't even bother to thread the reads at all but just to add the small bubbles to account for the alleles as they appear on the input VCF. Perhaps it actually composes the alternative haplotypes and does the threading as is done with the ref or the reads (describe bellow) but that would be inefficient if you have many variants in the active region ... why would you thread all the possible haplotypes when you can deal with each variant position individually?

Now come the reads... these are threaded starting a the first kmer that in unique (so is important to define what that actually means) and that is present already in the graph. So you may have a early kmer that is unique but because is not part of the reference sequence is not use as start.  I don't know the reason for this but I guess that authors (M. DePristo, R Poplin) wanted to have a connected graph for the very start. Anyway this process may result in **dangling tail** sequence that leave the reference path and never come back either because a suffix of the read that is not present in the reference sequence or it contains diverging non unique-kmers.

I think that after this is done the prefix of the read before such  a start kmer is thread upstream potentially resulting into a **dangling head**. 

Don't know what would happen with those sequence that don't have  any unique kmer that is not present in the reference.... or no unique kmer at all... are they simply discarded? possibly considering that this is typically unlikely expect for a read-long insertion and HC is not considered a good tool for those.

Also when we thread a read that has diverged from the reference and the next kmer is non-unique it will create a new node for that kmer even it is already present in the graph unless the it is already present as an extension of the previous kmer node.

Reply:
> "Up to this point, we have produced a de Bruijn graph in textbook fashion" is a very inaccurate statement as argued above.

See above - I'm trying to understand.

> I think that you mean to say twice or more

That's right.  I changed it to be clearer.

> I think it is important to explicitly state the concepts of unique and non-unique kmers

Agreed and done.

> The very first sequence that we thread is the reference sequence. In that special case we add all the kmers to the graph (I believe) unique or non-unique.

You're right.  Even after you told me it was easy to miss.  Thank you!

> I'm not sure what is done with the given alleles in GGA mode

I just checked and we use the GGA `VariantContexts` to create GGA haplotypes via substitution on ref allele --> alt allele in the ref haplotype.  Then we thread these as non-ref sequences after the ref haplotype and before the reads.  That is, we build GGA haplotypes and treat them as long reads.  I put this in the docs.

> So you may have a early kmer that is unique but because is not part of the reference sequence is not use as start

I *very hesitantly* think I got this right originally.  Let me try to make the case.  Looking at the code in `ReadThreadingGraph::threadSequence` it calls `findStart`, which goes through every kmer in the read and calls `isThreadingStart`, which is a one-liner:

`return startThreadingOnlyAtExistingVertex ? uniqueKmers.containsKey(kmer) : !nonUniqueKmers.contains(kmer)`

Since `startThreadingOnlyAtExistingVertex` is hard-coded to be `false`, a kmer is considered a start if it is *not non-unique*, even if hasn't been put on the graph yet.  After returning from `findStart`, we invoke `getOrCreateKmerVertex`, which creates a vertex for the kmer if one doesn't exist already.

> I think that after this is done the prefix of the read before such a start kmer is threaded upstream. . . Don't know what would happen with those sequence that don't have any unique kmer that is not present in the reference

If I'm right in the discussion above, these concerns would be moot.

Reply:
All what you saying seems reasonable.

Reply:
> Also when we thread a read that has diverged from the reference and the next kmer is non-unique it will create a new node for that kmer even it is already present in the graph unless the it is already present as an extension of the previous kmer node

Let's check this out in the code of `ReadThreadingGraph::extendChainByOne`.  First it looks at all outgoing edges from the previous kmer.  If one of these edges goes to the new kmer we just increment its multiplicity, as you said.  Then we call `getUniqueKmerVertex`, which checks whether this a unique kmer (i.e. a non-non-unique kmer that has already been added to the graph).  If it's unique, we add an edge to the already existing vertex.  Otherwise, that is, if it is non-unique *or* hasn't been added to the graph yet, we create a new vertex.  At the end of this long analysis, you are correct.  I suppose this is another place where we diverge from a vanilla de Bruijn graph.